### PR TITLE
Add resource check in xcatmn

### PR DESCRIPTION
**Pull request by xuweibj**
https://github.com/xcat2/xcat2-task-management/issues/62
* Enhance xcat deamon check
If more than 1 process for xcat deamon, fail
```
# xcatprobe xcatmn
[mn]: Checking all xCAT deamons are running...                                           [FAIL]
[mn]: More Deamon 'SSL listener' is running
[mn]: More Deamon 'DB Access' is running
```

* Check Linux ulimits configuration and network kernel parameter configuration on MN and SN.
Check xCAT deamon attributes configuration on MN.

Output of UT:
```
[mn]: Checking all xCAT deamons are running...                                                                    [ OK ]
[mn]: Checking xcatd can receive command request...                                                               [ OK ]
[c910f03c17k17]: Checking all xCAT deamons are running...                                                         [ OK ]
[mn]: Checking 'site' table is configured...                                                                      [ OK ]
[mn]: No interface provided by '-i' option, detected site table IP attribute 10.3.5.4, checking xCAT configurat...[WARN]
[c910f03c17k17]: Checking xcatd can receive command request...                                                    [ OK ]
[c910f03c17k17]: Checking 'site' table is configured...                                                           [ OK ]
[c910f03c17k17]: Checking provision network is configured...                                                      [ OK ]
[mn]: If this is incorrect, rerun with -i <ifname> option                                                         [WARN]
[mn]: Checking provision network is configured...                                                                 [ OK ]
[mn]: Checking 'passwd' table is configured...                                                                    [ OK ]
[mn]: Checking important directories(installdir,tftpdir) are configured...                                        [ OK ]
[mn]: Checking SELinux is disabled...                                                                             [ OK ]
[mn]: Checking HTTP service is configured...                                                                      [ OK ]
[mn]: Checking TFTP service is configured...                                                                      [ OK ]
[c910f03c17k17]: Checking 'passwd' table is configured...                                                         [ OK ]
[c910f03c17k17]: Checking important directories(installdir,tftpdir) are configured...                             [ OK ]
[mn]: Checking DNS service is configured...                                                                       [ OK ]
[c910f03c17k17]: Checking SELinux is disabled...                                                                  [ OK ]
[c910f03c17k17]: Checking HTTP service is configured...                                                           [ OK ]
[c910f03c17k17]: Checking TFTP service is configured...                                                           [ OK ]
[c910f03c17k17]: Checking DNS service is configured...                                                            [ OK ]
[c910f03c17k17]: Checking DHCP service is configured...                                                           [FAIL]
[c910f03c17k17]: There isn't 'dhcpd' deamon in current server
[c910f03c17k17]: Checking NTP service is configured...                                                            [FAIL]
[c910f03c17k17]: ntpd service is not running! Please setup ntp in current node
[c910f03c17k17]: Checking rsyslog service is configured...                                                        [ OK ]
[c910f03c17k17]: Checking firewall is disabled...                                                                 [ OK ]
[c910f03c17k17]: Checking minimum disk space for xCAT ['/var' needs 1GB;'/install' needs 10GB;'/tmp' needs 1GB]...[WARN]
[c910f03c17k17]: '/install',this directory is a part of file system '/install'. The free space available in dir...
*[c910f03c17k17]: Checking Linux ulimits configuration...                                                          [ OK ]
*[c910f03c17k17]: Checking network kernel parameter configuration...                                               [ OK ]
[mn]: Checking DHCP service is configured...                                                                      [ OK ]
[mn]: Checking NTP service is configured...                                                                       [ OK ]
[mn]: Checking rsyslog service is configured...                                                                   [ OK ]
[mn]: Checking firewall is disabled...                                                                            [ OK ]
[mn]: Checking minimum disk space for xCAT ['/var' needs 1GB;'/install' needs 10GB;'/tmp' needs 1GB]...           [WARN]
[mn]: '/var','/install','/tmp',these directories are parts of file system '/'. The free space available in dire...
*[mn]: Checking Linux ulimits configuration...                                                                     [ OK ]
*[mn]: Checking network kernel parameter configuration...                                                          [ OK ]
*[mn]: Checking xCAT deamon attributes configuration...                                                            [ OK ]
[mn]: Checking xCAT log is stored in /var/log/xcat/cluster.log...                                                 [ OK ]
[mn]: Checking xCAT management node IP: <10.3.5.4> is configured to static...                                     [ OK ]
[mn]: Checking dhcpd.leases file is less than 100M...                                                             [ OK ]
=================================== SUMMARY ====================================
[MN]: Checking on MN...                                                                                           [ OK ]
    No interface provided by '-i' option, detected site table IP attribute 10.3.5.4, checking xCAT configuratio...[WARN]
    If this is incorrect, rerun with -i <ifname> option                                                           [WARN]
    Checking minimum disk space for xCAT ['/var' needs 1GB;'/install' needs 10GB;'/tmp' needs 1GB]...             [WARN]
        '/var','/install','/tmp',these directories are parts of file system '/'. The free space available in di...
[SN:c910f03c17k17]: Checking on SN c910f03c17k17...                                                               [FAIL]
    Checking DHCP service is configured...                                                                        [FAIL]
        There isn't 'dhcpd' deamon in current server
    Checking NTP service is configured...                                                                         [FAIL]
        ntpd service is not running! Please setup ntp in current node
    Checking minimum disk space for xCAT ['/var' needs 1GB;'/install' needs 10GB;'/tmp' needs 1GB]...             [WARN]
        '/install',this directory is a part of file system '/install'. The free space available in directory '/...
```

If open file number over setting, fail 
```
[mn]: Checking Linux ulimits configuration...                                                                     [FAIL]
[mn]:The number of open files is not enough for xcatd service, increase the limits for it according to xCAT document
```
Over 80%, warning
```
[mn]: Checking Linux ulimits configuration...                                                                     [WARN]
[mn]: The number of open files is not enough for xcatd service, increase the limits for it according to xCAT document
```
If number of arp cache over setting, fail
``` 
[mn]: Checking network kernel parameter configuration...                                                          [FAIL]
[mn]: Most ARP has been used, please tuning network parameter as document
```
Over 80%, warning
```
[mn]: Checking network kernel parameter configuration...                                                          [WARN]
[mn]: Most ARP has been used, please tuning network parameter as document
```
If "xcatmaxconnections" is less than "xcatmaxbatchconnections", fail
```
[mn]: Checking xCAT deamon attributes configuration...                                                            [FAIL]
[mn]: Attribute xcatmaxbatchconnections must be less than xcatmaxconnections.
```
If number of nodes is over 500, and xcatmaxconnections or xcatmaxbatchconnections less than default, warning
```
[mn]: Checking xCAT deamon attributes configuration...                                                            [WARN]
[mn]: Management nodes are more than 500, please tuning xCAT daemon attributes as document
```

### CODE REVIEW

File: xCAT-probe/subcmds/xcatmn [hunk 1]
```diff
@@ -1042,6 +1062,104 @@ sub check_dhcp_leases {
...
+    my $error_ref      = shift;
+    my $rst            = 0;
+    my $rst_type       = "w";
+
+    $$checkpoint_ref = "Checking network kernel parameter configuration...";
+    @$error_ref = ();
+
+    my $net_set_file = "/etc/sysctl.conf";
+    my $net_gc_thresh = 512;
+    my $net_set_value = `cat $net_set_file | grep "net.ipv4.neigh.default.gc_thresh2" | awk -F'=' '{print \$2}' | tr '\n' ' ' | sed s/[[:space:]]//g`;
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/xcatmn]: Maybe can try
```
awk -F"=" '/net.ipv4.neigh.default.gc_thresh2/ {print $2}' /etc/sysctl.conf |sed s/[[:space:]]//g
```

File: xCAT-probe/subcmds/xcatmn [hunk 2]
```diff
@@ -284,7 +300,11 @@ sub check_all_xcat_deamons {
 
     my $output = `ps aux 2>&1|grep -v grep|grep xcatd`;
     foreach my $deamon (@deamon_list) {
-        if ($output !~ /$deamon/) {
+        my $counter = $output =~ s/$deamon/$deamon/g;
+        if ($counter > 1) {
+            push @$error_ref, "More Deamon '$deamon' is running";
```
Comment (robin2008) [inline xCAT-probe/subcmds/xcatmn]: I think it should be `Daemon`,  and here it should be some special handling for SSL and DB.
As it is normal when run `restartxcatd` and there are plugins not completed.  So you need to get the previous SSL pid and get all its children , and the give an reasonable judgement.

File: xCAT-probe/subcmds/xcatmn [hunk 3]
```diff
@@ -284,7 +300,11 @@ sub check_all_xcat_deamons {
...
+            $rst = 1;
+        } elsif ($counter == 0) {
             push @$error_ref, "Deamon '$deamon' isn't running";
```
Comment (robin2008) [inline xCAT-probe/subcmds/xcatmn]: typo for `deamon`

File: xCAT-probe/subcmds/xcatmn [hunk 4]
```diff
@@ -1042,6 +1062,104 @@ sub check_dhcp_leases {
...
+    my $rst            = 0;
+    my $rst_type       = "w";
+
+    $$checkpoint_ref = "Checking Linux ulimits configuration...";
+    @$error_ref = (); 
+
+    my $nofile_num = `sh -c 'ulimit -n' 2>&1`;
+    chomp($nofile_num);
+
+    my $process_id = `ps aux | grep "xcatd: SSL listener" | grep -v grep | awk -F' ' '{print \$2}'`;
```
Comment (robin2008) [inline xCAT-probe/subcmds/xcatmn]: Do you consider there are two SSL listener process?

File: xCAT-probe/subcmds/xcatmn [hunk 5]
```diff
@@ -1042,6 +1062,104 @@ sub check_dhcp_leases {
...
+    chomp($process_id);
+    my $process_folder = "/proc/$process_id/fd/";
+
+    my $open_num = 0;
+    $open_num = `ls $process_folder | wc -l` if (-e $process_folder);
+    chomp($open_num);
```
Comment (robin2008) [inline xCAT-probe/subcmds/xcatmn]: If the process folder has some issues, $open_num will be 0, you need to check it first and ignore the following.

File: xCAT-probe/subcmds/xcatmn [hunk 6]
```diff
@@ -1042,6 +1062,104 @@ sub check_dhcp_leases {
...
+
+    my $percent = $open_num/$nofile_num;
+    if ($percent >= 1) {
```
Comment (robin2008) [inline xCAT-probe/subcmds/xcatmn]: Better to use 
```
unless $percent < 0.8 { 
push @$error_ref, "The number of open files is not enough for xcatd service, increase the limits for it according to xCAT document";
$rst = 1;
$rst_type = "f" if ( $percent >= 1 );

```

File: xCAT-probe/subcmds/xcatmn [hunk 7]
```diff
@@ -267,30 +283,45 @@ sub print_check_result {
...
-            push @$error_ref, "Deamon '$deamon' isn't running";
+    foreach my $daemon (@daemon_list) {
+        my $counter = $output =~ s/$daemon/$daemon/g;
+        if ($counter > 1) {
+            if ($daemon == "SSL listener")  {
+                my $cur_pid = `cat /var/run/xcatd.pid`;
+                my @ssl_pids = `ps aux 2>&1|grep -v grep|grep "xcatd: $daemon"|awk -F' ' '{print \$2}'`; 
+                foreach my $ssl_pid (@ssl_pids) {
+                    next if ($cur_pid == $ssl_pid);
+                    my $child_pid = `pstree -p $ssl_pid 2>&1 | grep "SSL" | wc -l`;
```
Comment (robin2008) [inline xCAT-probe/subcmds/xcatmn]: Not sure `pstree` is always there, could we use `ps` to get the information?

**Pull request merged**